### ISSUE TITLE
Customized <myResource>.resw file name is not supported in Xaml(xUid)

### ISSUE THREAD

**Issue by sallyc3** [microsoft/WindowsAppSDK#1842]
### Discussed in https://github.com/microsoft/WindowsAppSDK/discussions/1841

<div type='discussions-op-text'>

<sup>Originally posted by **sallyc3** November 22, 2021</sup>
String resource in Strings/Resource.resw works for .xaml : x:Uid="myString"

String resource in Strings/MyResource.resw does not work for .xaml : xUid="/MyResource/myString".  This customized MyResource.resw resource file name works in UWP application.</div>

**Comment by sallyc3** [comment]
Strings/Resources.resw works, not Strings/Resource.resw.

**Comment by sallyc3** [comment]
Please help.  We need localization solution using non Resources file name (Resources.resw)) resource files for WinUI 3.0.

<myResource>.resw does not work in cpp either with 
Windows::ApplicationModel::Resources::ResourceLoader::GetForCurrentView();   => Resources.resw
nor with
Windows::ApplicationModel::Resources::ResourceLoader::GetForCurrentView(L"myWRC/MyResource");  => MyResource.resw

Does not work with following sequence.  After GetForCurrentView() is called, it dropped to the end of the procedure.  Seems some thread issue(?).
auto resourceLoader = Windows::ApplicationModel::Resources::ResourceLoader::GetForCurrentView() ;
<textBlockitem>.Name(resourceLoader.GetString(L"myString"));

**Comment by andrewleader** [comment]
Hey @axelandrejs @huichen123, would either of you know whether XAML's x:Uid can work with Resource files that aren't explicitly named `Resources.resw`? It sounds like if you rename the file to `MyResources.resw` or anything else.

**Comment by andrewleader** [comment]
Also @sallyc3, to set expectations, it's Thanksgiving week here and a lot of folks are out of the office, so you might not get an answer till next week.

**Comment by sallyc3** [comment]
I understand.  As long someone will get back to me with an answer next week - how to use <myResource>.resw in both xaml and cpp.  Thanks.

**Comment by huichen123** [comment]
@evelynwu-msft It seems WinUI treats x:Uid as a submap under Resources. Is that expected?

**Comment by evelynwu-msft** [comment]
Yes, we have always expected string resources used for x:Uid to be under the `Resources/` submap.

**Comment by huichen123** [comment]
@evelynwu-msft is this unique to WinUI3? @sallyc3 mentioned it works in UWP.

**Comment by sallyc3** [comment]
I followed the instruction in the following link and it worked for UWP.
https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest#factoring-strings-into-multiple-resources-files

       Thanks.
-Sally

**Comment by evelynwu-msft** [comment]
@huichen123 The way WinUI calls into MRT Core to resolve the x:Uid is a little different from how UWP XAML calls into system MRT; assuming that the PRI is still being constructed in the same fashion, then that means a subtle bug was introduced during the port.

**Comment by sallyc3** [comment]
Please also address the CPP implementation issue.  It does not work from the following sample code from the same link<https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest#factoring-strings-into-multiple-resources-files> example.

// resource file is at Strings\en-US\ErrorMessages.resw
auto resourceLoader{ Windows::ApplicationModel::Resources::ResourceLoader::GetForCurrentView(L"ErrorMessages") };
myXAMLTextBlockElement().Text(resourceLoader.GetString(L"MismatchedPasswords"));

       Thanks.
-Sally

**Comment by huichen123** [comment]
The PRI construction didn't change. I am assigning this to @evelynwu-msft 

@sallyc3 as for your other issue in your last reply, does it work for resources in resources.resw? Since you are using system MRT, what is your OS version?

**Comment by sallyc3** [comment]
1.  as for your other issue in your last reply, does it work for resources in resources.resw?

**No, for CPP, it does not work even trying to use the Resource.resw.  Code sample :

    auto resourceLoader = Windows::ApplicationModel::Resources::ResourceLoader::GetForCurrentView() ;

item.Name(resourceLoader.GetString(L”myString”));

   Nothing happened.  The string for Name property is not displayed.  In debugger, after the GetString() call, it just went to the end of the current  routine.  Looks like a tread related issue(?).

** For Xaml, Resouce.resw works with x:uid=”MyString”



  1.  Since you are using system MRT, what is your OS version?

**Edition             Windows 10 Enterprise

Version 21H1

Installed on         ‎5/‎24/‎2021

OS build               19043.1348

Experience         Windows Feature Experience Pack 120.2212.3920.0



       Thanks.

-Sally

**Comment by huichen123** [comment]
you may run on non-UI thread? Try GetForViewIndependentUse instead.

**Comment by sallyc3** [comment]
Ok.  Will give that a try.
      Thanks.
-Sally

**Comment by sallyc3** [comment]
GetForViewIndependentUse() does not work.
It fixed the treading issue.  But still not getting the string from Resources.resw.
Code sample :

auto resourceLoader = Windows::ApplicationModel::Resources::ResourceLoader::GetForViewIndependentUse();
item.Name(resourceLoader.GetString(L"LoadingStr"));
// item.Name == L""

-Sally

**Comment by huichen123** [comment]
can you dump resources.pri with makepri.exe to see if there is a Resources\MyString entry?

**Comment by sallyc3** [comment]
Note, the resources are from the .dlls.  So the pri files is not resources.pri.
The problem could be that CPP is only looking into resources.pri? not <dll file>.pri?

I also tried to change the Region to Australia, the localization does not work for both Xaml and CPP. It does not take the strings from the Strings\en-au\Resources.resw file.
Localization works with UWP, for both Xaml and CPP.

Pri file for AccessServerWRC :  x64\Debug\AccessServerWRC\AccessServerWRC.pri

makepri.exe dump of AccessServerWRC.pri to .xml file. The xml file shows correctly "EmptyProjectStr" for both en-us and en-au (Australia).

<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PriInfo>
                <ResourceMap name="AccessServerWRC" version="1.0" primary="true">
                                <Qualifiers>
                                                <Language>EN-US,EN-AU</Language>
                                </Qualifiers>
                                <ResourceMapSubtree name="AccessServerWRC">
                                                <ResourceMapSubtree name="AccessPageRes">
                                                                <NamedResource name="EmptyProjectStr" uri="ms-resource://AccessServerWRC/AccessServerWRC/AccessPageRes/EmptyProjectStr">
                                                                                <Candidate qualifiers="Language-EN-US" isDefault="true" type="String">
                                                                                                <Value>Empty Project</Value>
                                                                                </Candidate>
                                                                                <Candidate qualifiers="Language-EN-AU" isDefault="true" type="String">
                                                                                                <Value>~~Empty Project</Value>
                                                                                </Candidate>
                                 ...

**Comment by sallyc3** [comment]
It works with "AccessServerWRC/Resources", not "\".  So this is good.
I can also get different .resw file name working with "AccessServerWRC/myResources".
But it does not work for localization with different country region setting.  Please help in addressing the localization issue.  When changed Region to Austrilia, it does not use Strings\en-au\Resources.resw for example.  This works in UWP.

I found the resources.pri in the application bin\x64\Debug\resources.pri
Using the name from the resources.pri dump, both xaml and CPP works with both Resources.resw and MyResources.resw.
xaml - x:Uid = "/MyWRC/MyResources/MyString"
CPP - 
auto resourceLoader = Windows::ApplicationModel::Resources::ResourceLoader::GetForViewIndependentUse(L"MyWRC/MyResources");
item.Name(resourceLoader.GetString(L"MyString"));
Thanks!

**Comment by huichen123** [comment]
is your app packaged or unpackaged?

**Comment by sallyc3** [comment]
Not sure.  I think is unpackaged.  Could not get packaged working.

Has both /App/App and /App_WAP/App (Package) directories.

Source code are in /App/App

/App_WAP/App (Package)/bin/x64/Debug/resources.pri

-Sally

**Comment by riverar** [comment]
I think to better help you @sallyc3 we need a minimal reproducible example. Create a sample project with same resources, put it on GitHub or zip it up and provide a link. This back and forth with tidbits of information is not really productive. All those fields you skipped in the bug report form are really required.

**Comment by sallyc3** [comment]
Actually Hui helped me to solve the xaml and cpp issues of using myResoueces.resw. So it’s productive. I got what I need. 

As for the new issue I found out about the localization, Hui asked if I’m using packaged app or not as her last question. I answered. I’d like to wait till she gets back to me on the suggestion 1st. 

       Thanks. 

-Sally

**Comment by riverar** [comment]
@sallyc3 Yes, and your answer was "I think is unpackaged" which is frustratingly unhelpful. This thread is frankly full of back and forth that could have been avoided had you simply completed the issue template and provided a repro case. Perhaps it's too late now, but in future issues, please consider completing the template and providing a repro as expected in most open source communities.

**Comment by sallyc3** [comment]
Sure. I understand. 

-Sally

**Comment by huichen123** [comment]
You are using system MRT. For unpackaged app, it respects Windows display language instead of user preferred languages. To let user preferred languages take effect, you may have to switch to MRTCore (https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources?view=windows-app-sdk-1.0)

**Comment by sallyc3** [comment]
We will try to make the app as packaged again now that Windows AppSDK 1.0 is released.  Hope things are working better.  We had issues with packaged app before.

Will let you know of the result.

       Thanks.
-Sally

**Comment by sallyc3** [comment]
Hi Hui,

Thank you so much for all your support.  Everything is working with the resource string table and localization.  Our app package works with localization now.  I probably set the incorrect resource loader name yesterday.

To summarize what I need to do for the WindowsAppSDK in using our own MyRes.resw file name of the dll's.  This is different from how we did for UWP.
For Xaml
x:Uid="/MyWRC/MyRes/MyString"

For CPP
#include <winrt/Windows.ApplicationModel.Resources.h>
auto resourceLoader = Windows::ApplicationModel::Resources::ResourceLoader::GetForViewIndependentUse(L"MyWRC1/MyRes1");
winrt::hstring str = resourceLoader.GetString(L"MyString1");


I also learned to use makepri.exe to dump out the resource strings defined in the app for references

makepri.exe dump /if .\MyWRC.pri /of .\resources.pri.xml

I know WindowsAppSDK is very new and we just updated to the 1.0 NuGet package from the preveiw3 version.  I am looking forward for the full production of WindowsAppSDK with more thorough documentation.

      Thanks again for your prompt and great support!

-Sally

**Comment by sallyc3** [comment]
Also many thanks to EvelynWu of her prompt technical support!
-Sally

**Comment by huichen123** [comment]
I am glad you have your problem resolved. I am reopening this issue and assign to @evelynwu-msft as there is still an issue in WinUI3 how x:Uid is handled.

**Comment by sallyc3** [comment]
Ok. Thanks. 

-Sally

**Comment by sallyc3** [comment]
I use packaged app.
Still have problem in using de German string resources in setting the Windows display language to German.
en-au Australia worked before then I thought it will work for all non en-US.  Since Australia is also a version of English so it worked for all the English languages.  Once changed to non-English display language, then the resw does not work for both Xaml and CPP.  Please help.

**Comment by sallyc3** [comment]
What template do I use for packaged and unpackaged WinUI 3 in Desktop?

Blank App, Packaged (WinUI 3 in Desktop) : can not use this one.  Navigation does not work.  Filed an issue on this already.
Blank App, Packaged with WAP (WinUI 3 in Desktop) : We are using this one now.  Localization does not work for non-English resource files.
Windows Runtime Component (WinUI 3) : this is used for component only.

How do you create unpackaged app?

**Comment by andrewleader** [comment]
@sallyc3 can you share the link to your navigation issue?

To create an unpackaged WinUI 3 app, you have to use the _Blank App, Packaged (WinUI 3 in Desktop)_ template.

**Comment by sallyc3** [comment]
This one is the Navigation one.
Migrating UWP app to WindowsAppSDK. Can't navigate to page from Windows Runtime Component · Issue #1769 · microsoft/WindowsAppSDK (github.com)

This one has to do with win32 not being available in runtime components
Win32 Desktop functions not available in WinUI 3 Runtime Component. · Issue #1780 · microsoft/WindowsAppSDK (github.com)

**Comment by huichen123** [comment]
@sallyc3 when you use MRTCore, Windows Display Language is irrelevant. User preferred languages are used.

**Comment by sallyc3** [comment]
Our app is unpackaged, explained from https://github.com/microsoft/WindowsAppSDK/issues/1842#issuecomment-992735657

I did not change the code to use MRTCore.  I like to follow Windows Display Language, set in Regions->Language->Windows display language.
In our unpackaged app, Windows Display Language setting (to de German for example) does not work.  But if I change the Default Language in the app Package.appxmanifest, then the default language (de for example) works for my app.  I did not need to change any code or resource file locations to get package default language working.
But we need to give the user control to set their own Windows Display Language to see the their Windows Display language showing in our app GUI.

The internationalization of the locale support is still confusing.  I hope there is a clear instruction in supporting this in WinUI3.0.
Similar to the instruction for UWP at https://docs.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest#factoring-strings-into-multiple-resources-files
